### ISSUE TITLE
require-atomic-updates complains about unrelated variable references

### ISSUE THREAD

**Issue by dohse** [eslint/eslint#11900]
**Tell us about your environment**

* **ESLint Version: 6.0.1**
* **Node Version: 10.15.3**
* **npm Version: 6.4.1**

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```yaml
---
  extends: eslint:recommended
  parserOptions:
    ecmaVersion: 2017
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
async function f() {}

async function t() {
  for (const x of ['A', 'B', 'C']) {
    await f(x);
  }
  for (const x of ['A', 'B', 'C']) {
    await (() => f(x))();
  }
}

t();
```

```bash
> eslint atomic-updates.js
```

**What did you expect to happen?**

I get no error or a useful explanation what is wrong with this code.

**What actually happened? Please include the actual, raw output from ESLint.**

I am getting this linting error:

```
…/atomic-updates/atomic-updates.js
  7:3  error  Possible race condition: `const x` might be reassigned based on an outdated value of `const x`  require-atomic-updates

✖ 1 problem (1 error, 0 warnings)
```

**Are you willing to submit a pull request to fix this bug?**

I expect it to take me quiet long to understand the logic. Therefore not at this moment.

**Comment by hon2a** [comment]
I've just encountered a similar one:

```javascript
try {
  result = await someAsyncCall(/* ... */)
} catch (e) {
  ctx.status = e.statusCode || 500
  ctx.body = e.error
  return
}
```

I'm getting an error for both of the assignment lines in the `catch` block:

```
[error] Possible race condition: `ctx.status` might be reassigned based on an outdated value of `ctx.status`. (require-atomic-updates)
[error] Possible race condition: `ctx.body` might be reassigned based on an outdated value of `ctx.body`. (require-atomic-updates)
```

**Comment by mysticatea** [comment]
Thank you for your report.

I confirmed it. Looks like the initializer in `for-of` syntax was handled as an assignment.

@hon2a Your case looks different to this issue because it doesn't contain `for-of` syntax. And I couldn't reproduce it: [online demo](https://eslint.org/demo/#eyJ0ZXh0IjoiYXN5bmMgZnVuY3Rpb24gZigpIHtcclxuXHJcbnRyeSB7XHJcbiAgcmVzdWx0ID0gYXdhaXQgc29tZUFzeW5jQ2FsbCgvKiAuLi4gKi8pXHJcbn0gY2F0Y2ggKGUpIHtcclxuICBjdHguc3RhdHVzID0gZS5zdGF0dXNDb2RlIHx8IDUwMFxyXG4gIGN0eC5ib2R5ID0gZS5lcnJvclxyXG4gIHJldHVyblxyXG59XHJcblxyXG59XHJcbi8qZXNsaW50IHJlcXVpcmUtYXRvbWljLXVwZGF0ZXM6IGVycm9yICovIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo4LCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19).

**Comment by hon2a** [comment]
@mysticatea: [Sorry, here you go.](https://eslint.org/demo/#eyJ0ZXh0IjoiYXN5bmMgZnVuY3Rpb24gZigpIHtcclxuICBjb25zdCBzZXJ2aWNlID0gZ2V0U29tZXRoaW5nVXNpbmdDb250ZXh0KGN0eC5zb21lUHJvcClcclxuXHJcbiAgdHJ5IHtcclxuICAgIHJlc3VsdCA9IGF3YWl0IHNvbWVBc3luY0NhbGwoKVxyXG4gIH0gY2F0Y2ggKGUpIHtcclxuICAgIGN0eC5zdGF0dXMgPSBlLnN0YXR1c0NvZGUgfHwgNTAwXHJcbiAgICBjdHguYm9keSA9IGUuZXJyb3JcclxuICAgIHJldHVyblxyXG4gIH1cclxuXHJcbiAgY3R4LmJvZHkgPSByZXN1bHQuYm9keVxyXG4gIGN0eC5zdGF0dXMgPSByZXN1bHQuc3RhdHVzQ29kZVxyXG59XHJcbi8qZXNsaW50IHJlcXVpcmUtYXRvbWljLXVwZGF0ZXM6IGVycm9yICovIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo4LCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19) Should I create a separate issue then?

**Comment by mysticatea** [comment]
@hon2a Thanks. It looks the same issue as #11899.

**Comment by nicolo-ribaudo** [comment]
Another demo case: [demo](https://eslint.org/demo#eyJ0ZXh0IjoiZXhwb3J0IGxldCBsb2NrID0gMjtcblxuZXhwb3J0IGFzeW5jIGZ1bmN0aW9uIGYoKSB7XG4gIHtcbiAgICBsZXQgbG9jaztcbiAgICBsb2NrO1xuICB9XG4gIGF3YWl0IG51bGw7XG4gIGxvY2sgPSAzO1xufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTAsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWtleXMiOjIsIm5vLWR1cGxpY2F0ZS1jYXNlIjoyLCJuby1lbXB0eSI6Miwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1lbXB0eS1wYXR0ZXJuIjoyLCJuby1leC1hc3NpZ24iOjIsIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6Miwibm8tZXh0cmEtc2VtaSI6Miwibm8tZmFsbHRocm91Z2giOjIsIm5vLWZ1bmMtYXNzaWduIjoyLCJuby1nbG9iYWwtYXNzaWduIjoyLCJuby1pbm5lci1kZWNsYXJhdGlvbnMiOjIsIm5vLWludmFsaWQtcmVnZXhwIjoyLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6Miwibm8tbWlzbGVhZGluZy1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcHJvdG90eXBlLWJ1aWx0aW5zIjoyLCJuby1yZWRlY2xhcmUiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtY2F0Y2giOjIsIm5vLXVzZWxlc3MtZXNjYXBlIjoyLCJuby13aXRoIjoyLCJyZXF1aXJlLWF0b21pYy11cGRhdGVzIjoyLCJyZXF1aXJlLXlpZWxkIjoyLCJ1c2UtaXNuYW4iOjIsInZhbGlkLXR5cGVvZiI6Mn0sImVudiI6e319fQ==)

```js
export let lock = 2;

export async function f() {
  {
    let lock;
    lock;
  }
  await null;
  lock = 3;
}
```
